### PR TITLE
add guide on how to review infra changes as transactions

### PR DESCRIPTION
**Pull request by depombo**

### CODE REVIEW

File: site/docs/how-to/review.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+- works with IaSQL schema upgrades
+- works with any version control system, not just GitHub, without the need for privileged access
+- preserves the ability to `iasql_preview` changes in development and get the type-safety feedback from the dashboard editor
+
+```sql title="Create review for an infrastructure change within a transaction"
+SELECT * FROM iasql_begin();
+
+UPDATE log_group SET log_group_name = 'test' WHERE id = 4;
+
+SELECT * FROM iasql_create_review('My review' )
```
Comment (aguillenv) [inline site/docs/how-to/review.md:21]:
```suggestion
SELECT * FROM iasql_create_review('My review')
```

Comment (aguillenv) [reply site/docs/how-to/review.md:21]: should we add in this example the `description` too so users can know how to use it?

File: site/docs/how-to/review.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+```
+# My review
+
+Review description
+
+## IaSQL Preview
+
+| action | table_name | id | description |
+| --- | --- | --- | --- |
+| update | log_group | 4 | quickstart-log-group|us-east-2 |
```
Comment (aguillenv) [inline site/docs/how-to/review.md:35]:
```suggestion
| update | log_group | 4 | quickstart-log-group | us-east-2 |
```

File: site/docs/how-to/review.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+We set forth the best convention to review infrastructure changes within transactions which:
+- works with any cloud identity and access management structure
+- works with IaSQL schema upgrades
+- works with any version control system, not just GitHub, without the need for privileged access
+- preserves the ability to `iasql_preview` changes in development and get the type-safety feedback from the dashboard editor
+
+```sql title="Create review for an infrastructure change within a transaction"
+SELECT * FROM iasql_begin();
+
+UPDATE log_group SET log_group_name = 'test' WHERE id = 4;
```
Comment (aguillenv) [inline site/docs/how-to/review.md:19]:
```suggestion
UPDATE log_group SET log_group_arn = 'test' WHERE log_group_name = 'quickstart-log-group';
```

File: site/docs/how-to/review.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+# Change log group name
+
+Review description
+
+## IaSQL Preview
+
+| action | table_name | id | description |
+| --- | --- | --- | --- |
+| update | log_group | 4 | quickstart-log-group|us-east-2 |
```
Comment (aguillenv) [inline site/docs/how-to/review.md:69]:
```suggestion
| update | log_group | 4 | quickstart-log-group | us-east-2 |
```

File: site/docs/how-to/review.md (hunk 1 above)
Comment (dfellis) [reply site/docs/how-to/review.md:21]: I agree with @aguillenv that this doc should have the description column filled in, but maybe mention that it's optional?

File: site/docs/how-to/review.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+---
+sidebar_position: 5
+slug: '/review'
+---
+
+# Peer review infrastructure changes
+
+Eng teams move over from using cloud UIs to IaC because cloud infrastructure changes can lead to outages more often than business logic changes. As a result, it is considered a DevOps best practice to code review infrastructure changes using IaC tools and show the `preview` of the resulting change in pull requests to raise the quality bar and reduce the possibility of mistakes. We do not recommend that cloud infrastructure changes done with IaSQL be code-reviewed using migration systems since the IaSQL schema is defined by [modules](../concepts/module.md) that can be upgraded and changed over time. Furthermore, we recommend complex, or delicate, infrastructure changes should be treated as transactions in IaSQL. [IaSQL transactions](../concepts/transaction.md) are a bit different from normal DB transactions in that they are multiuser and can be previewed with the [`iasql_preview`](../modules/builtin/iasql_functions.md) function for the changes that the ongoing transaction would create in the cloud, similar to Pulumi’s `preview` or Terraform’s `plan`.
```
Comment (dfellis) [inline site/docs/how-to/review.md:8]: Restructuring things a bit to be clearer, and breaking up this paragraph into a few for better emphasis of the various points.

```suggestion
Engineering teams move from using cloud UIs to IaC because cloud infrastructure changes can lead to outages more often than business logic changes. As a result, it is a DevOps best practice to code review infrastructure changes using IaC tools, showing the `preview` of the resulting change in pull requests to raise the quality bar and reduce the possibility of mistakes.

The standard practice for code review of database changes is to write the changes via a [migration system](https://en.wikipedia.org/wiki/Schema_migration) that are executed after approval. We do not recommend that cloud infrastructure changes done with IaSQL follow this pattern for two reason: first, these migration tools often do not support data changes, only schema changes, and the IaSQL schema is defined by [modules](../concepts/module.md) that can be upgraded and changed over time, and second, you lose all of the feedback the database can give you on whether or not your SQL statements are valid and what impact they will have on your cloud infrastructure.

Instead, we recommend complex, or delicate, infrastructure changes should be treated as transactions in IaSQL. [IaSQL transactions](../concepts/transaction.md) are a bit different from normal DB transactions in that they lock all database changes from propagating into the cloud but do not lock the changes from being viewed by other database connections, making them "multiuser", so others can inspect the changes, and can preview what they would do with the [`iasql_preview`](../modules/builtin/iasql_functions.md) function, similar to Pulumi’s `preview` or Terraform’s `plan`.
```

File: site/docs/how-to/review.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+We set forth the best convention to review infrastructure changes within transactions which:
```
Comment (dfellis) [inline site/docs/how-to/review.md:10]:
```suggestion
But, reviewable artifacts are still very useful to summarize and explain the intent behind a proposed set of changes, so we have created an easy-to-follow yet comprehensive convention to review infrastructure changes within transactions which:
```

File: site/docs/how-to/review.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+- works with any version control system, not just GitHub, without the need for privileged access
+- preserves the ability to `iasql_preview` changes in development and get the type-safety feedback from the dashboard editor
+
+```sql title="Create review for an infrastructure change within a transaction"
+SELECT * FROM iasql_begin();
+
+UPDATE log_group SET log_group_name = 'test' WHERE id = 4;
+
+SELECT * FROM iasql_create_review('My review' )
+```
```
Comment (dfellis) [inline site/docs/how-to/review.md]: I think we should add some SQL comments in this block to explain the parts of this process:

```suggestion
```sql title="Create review for an infrastructure change within a transaction"
-- First, we open a new transaction with `iasql_begin`
SELECT * FROM iasql_begin();

-- Then we make the infrastructure changes we need.
-- If a SQL statement is invalid, Postgres will prevent it
-- from running and it won't end up in the transaction, so
-- we can use it as our IDE.
UPDATE log_group SET log_group_name = 'test' WHERE id = 4;

-- We create the review artifact with `iasql_create_review`, providing
-- a title and a description of why we're doing this.
SELECT * FROM iasql_create_review('My infra change', 'Why this change is necessary')
```
```

File: site/docs/how-to/review.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+The raw markdown content returned will look as follows:
```
Comment (dfellis) [inline site/docs/how-to/review.md:24]:
```suggestion
The output of `iasql_create_review` is markdown-formatted text and will look as follows:
```

File: site/docs/how-to/review.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+```
+# My review
+
+Review description
```
Comment (dfellis) [inline site/docs/how-to/review.md:29]:
```suggestion
# My infra change

Why this change is necessary
```

File: site/docs/how-to/review.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+## SQL changes
+
+```sql
+
+UPDATE log_group
+SET log_group_name = 'quickstart-log-group', log_group_arn = 'test', creation_time = '2023-01-25T11:17:14.045+00:00', region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')
+WHERE log_group_name = 'quickstart-log-group' AND log_group_arn = 'arn:aws:logs:us-east-2:257682470237:log-group:quickstart-log-group:*' AND creation_time = '2023-01-25T11:17:14.045+00:00' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2');
+
+```
```
Comment (dfellis) [inline site/docs/how-to/review.md]: I would move this up to before the `iasql_commit` call, perhaps saying something like "if you put this into a Github PR or other tool that pretty-prints markdown, it will look like this"

Comment (depombo) [reply site/docs/how-to/review.md]: I disagree. Rather, I did that initially and it becomes really hard to follow what is the example versus the actual doc so I would like to keep it here

File: site/docs/how-to/review.md [hunk 11]
```diff
@@ -5,34 +5,42 @@ slug: '/review'
...
 
-Review description
+Why this change is necessary
 
 ## IaSQL Preview
 
 | action | table_name | id | description |
 | --- | --- | --- | --- |
-| update | log_group | 4 | quickstart-log-group|us-east-2 |
+| update | log_group | 4 | quickstart-log-group | us-east-2 |
```
Comment (dfellis) [inline site/docs/how-to/review.md:43]: Btw, why does this have the log group name and the region? There are 5 columns here vs the 4 titled columns.

I think the fix @aguillenv suggested was kinda incorrect: I bet the `log_group` description generation was creating the string `quickstart-log-group|us-east-2` and that is playing havoc with the markdown table generation here.

I think the proper solution is to escape the `|` symbol in the text inserted into the tables and then show the escaped version here, instead.

**Review by yrobla — APPROVED** [review]

File: site/docs/how-to/review.md (hunk 11 above)
Comment (aguillenv) [reply site/docs/how-to/review.md:43]: Oh, you are right, I did not realize that in my PR :/

Comment (aguillenv) [reply site/docs/how-to/review.md:43]: This is being handled at https://github.com/iasql/iasql/pull/2349. It produces the following markdown for this case:

# New Review

Review description

## IaSQL Preview

| action | table_name | id | description |
| --- | --- | --- | --- |
| update | log_group | 2 | /aws/codebuild/oow1f-ecr-builder\|us-east-2 |

## SQL changes

```sql

UPDATE log_group
SET log_group_name = '/aws/codebuild/oow1f-ecr-builder', log_group_arn = 'tets', creation_time = '2023-01-25T12:43:59.107+00:00', region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')
WHERE log_group_name = '/aws/codebuild/oow1f-ecr-builder' AND log_group_arn = 'arn:aws:logs:us-east-2:257682470237:log-group:/aws/codebuild/oow1f-ecr-builder:*' AND creation_time = '2023-01-25T12:43:59.107+00:00' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2');

```

Comment (dfellis) [reply site/docs/how-to/review.md:43]: @depombo let's update the docs now that #2349 has landed :)

Comment (depombo) [reply site/docs/how-to/review.md]: good catch, done :)

File: site/docs/how-to/review.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+SELECT * FROM iasql_begin();
+-- Then we make the infrastructure changes we need.
+-- If a SQL statement is invalid, Postgres will prevent it
+-- from running and it won't end up in the transaction, so
+-- we can use it as our IDE.
+UPDATE log_group SET log_group_name = 'test' WHERE id = 4;
+-- We create the review artifact with `iasql_create_review`, providing
+-- a title and a description of why we're doing this.
+SELECT * FROM iasql_create_review('My infra change', 'Why this change is necessary')
+
```
Comment (dfellis) [inline site/docs/how-to/review.md:31]: The output looks weird when I preview it:

Screenshot from 2023-03-23 16-12-08

And I realized it's because you didn't close the code example here

````suggestion
```

````

File: site/docs/how-to/review.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+# My infra change
+
+Why this change is necessary
+
+## IaSQL Preview
+
+| action | table_name | id | description |
+| --- | --- | --- | --- |
+| update | log_group | 4 | quickstart-log-group\|us-east-2 |
+
```
Comment (dfellis) [inline site/docs/how-to/review.md:44]: Oh, also missing the closure here:

````suggestion
```

````

Comment (dfellis) [reply site/docs/how-to/review.md:45]: That explains why it didn't get *completely* screwed up in formatting.

File: site/docs/how-to/review.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+## SQL changes
+
+```sql
+
+UPDATE log_group
+SET log_group_name = 'quickstart-log-group', log_group_arn = 'test', creation_time = '2023-01-25T11:17:14.045+00:00', region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')
+WHERE log_group_name = 'quickstart-log-group' AND log_group_arn = 'arn:aws:logs:us-east-2:257682470237:log-group:quickstart-log-group:*' AND creation_time = '2023-01-25T11:17:14.045+00:00' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2');
+
+```
```
Comment (dfellis) [inline site/docs/how-to/review.md:55]: Sorry, I messed up, let me just fix that formatting

`````suggestion
````
# My infra change

Why this change is necessary

## IaSQL Preview

| action | table_name | id | description |
| --- | --- | --- | --- |
| update | log_group | 4 | quickstart-log-group\|us-east-2 |

## SQL changes

```sql

UPDATE log_group
SET log_group_name = 'quickstart-log-group', log_group_arn = 'test', creation_time = '2023-01-25T11:17:14.045+00:00', region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')
WHERE log_group_name = 'quickstart-log-group' AND log_group_arn = 'arn:aws:logs:us-east-2:257682470237:log-group:quickstart-log-group:*' AND creation_time = '2023-01-25T11:17:14.045+00:00' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2');
```
````
`````

**Comment by depombo** [comment]
Screenshot 2023-03-23 at 5 38 45 PM
how it looks on docusaurus

**Review by dfellis — APPROVED** [review]
